### ISSUE TITLE
GUI CPU usage

### ISSUE THREAD

**Issue by cliffordwolf** [YosysHQ/nextpnr#24]
Just running `../nextpnr-ice40 --hx8k --gui` (without loading a design or doing anything) already has a pretty high CPU usage.

**Comment by q3k** [comment]
@cliffordwolf out-of-band you told me it's especially bad on Nvidia/Linux. Would you mind providing an apitrace so I can look at what's the performance bottleneck in that case?

**Comment by cliffordwolf** [comment]
apitrace: http://scratch.clifford.at/nextpnr-ice40.trace.xz

**Comment by datenwolf** [comment]
@cliffordwolf since I told you OOB that I'd like to pull my weight on graphics/OpenGL issues, I've started hacking on this stuff.

There's a lot that can be improved in the OpenGL code performance wise: Lot of redundant copies, that can be avoided, picking could be performed through a ID buffer populated by the GPU while drawing (picking reduces to indexing into the generated buffer, reading the ID of the element therein). Selecting the drawing style (hover, grayed out, etc.) can be done entirely in the shader, changing the semantics of the graphics element list to allow for direct access by the GPU instead of requiring a coalesce copy each frame redraw…

**Comment by q3k** [comment]
Absolutely, the current renderer code can be improved quite a lot - it grew organically, and the first iteration of the shader code had quite a different scope in mind than now.

I don't currently have time to work on this, so go ahead :).

**Comment by datenwolf** [comment]
Already on it. Just want to make sure, we don't step on each others' toes.

**Comment by smunaut** [comment]
Any update on this ?  Starting an ECP5 GUI makes makes my fan spin ... it not that much CPU (not that I can see in `top` at least) I think it's a lot of wake ups / small cycles ?  Even when it's running on another virtual desktop it slows down my system.

**Comment by mmicko** [comment]
@smunaut There was no updates to GUI code coming from outside contributors so I will work on solving this issue. There is a lot of space for improvement, and also ECP5 is not that very large so performance issues should better be solved until we start supporting some larger FPGA.

**Comment by datenwolf** [comment]
I had to reprioritize my health (i.e. spend more time away from screens and with human beings) over adding that to my workload. Maybe I'll look at this again, sometime again.

**Comment by mmicko** [comment]
@datenwolf no worries there will work on this issue. There will always be place for new ideas and optimizations :)
